datasketches lib based aggregation and post-aggregation for quantiles/histograms

himanshug opened the pull request:
in a contrib extension

pdeva commented:
weird that this is failing on just jdk 7

xvrl commented on processing/src/test/java/io/druid/query/aggregation/AggregationTestHelper.java:
```diff
@@ -196,10 +208,22 @@ public static final AggregationTestHelper createSelectQueryAggregationTestHelper
         QueryRunnerTestHelper.NOOP_QUERYWATCHER
     );
 
+    IndexIO indexIO = new IndexIO(
```
we can probably reuse indexIO here like we do in other test helpers

himanshug replied:
It depends on specific ObjectMapper which gets appropriate Modules from Aggregator impl registered as provided by the user of this class. that is the reason it is not reused.

gianm commented on extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/QuantilesSketchAggregatorFactory.java:
```diff
@@ -0,0 +1,327 @@
+  {
+    // off-heap implementation of quantiles sketch is not available right now and hence
+    // nothing is stored off-heap really.
+    return 1;
```
maybe this should be something other than 1? I think using 1 here will make the query engines try to create a large number of bufferaggregators, and this could cause a lot of heap pressure.

himanshug replied:
not sure what you mean, since we are not storing anything on the BB so there is no reason to say we're using space there. 1 is minimum that can be used. number of BufferedAggregator wouldn't depend on this.

gianm replied:
sorry, I didn't mean creating a lot of bufferaggregators, I meant creating a lot of on-heap aggregators.

The situation I'm thinking of is doing a topN with a sketch as the metric. IIRC the topN would figure out how many rows to do per pass by adding up all the max intermediate sizes and dividing that into the size of the offheap buffer. So if the max intermediate size is 1 then this could lead to topN creating a large number of on-heap aggregators for a single pass.

himanshug replied:
@gianm topN itself should be able to handle that instead of this aggregator falsely reserving the off heap buffer capacity its not going to use, https://github.com/druid-io/druid/pull/2827 should address this.

cheddar replied:
I'm with @gianm on this one.  The implementation here breaks the current model that all other aggregators use and if you want to use it, you now have to tune your heap size differently than you previously had to, and you only have to tune it for this one query.  The "fix" in #2827 will end up enforcing some global limits on how much work can be done on any and every query, even when that limitation doesn't make sense.

The semantics inside of Druid is for aggregators to specify how much memory they use and for query implementations to leverage that.  One of the following things should happen

1. Make the implementation off-heap (continuously serialize and deserialize the object over and over again)
2. Throw an `UnsupportedOperationException` from `factorizeBuffered` and hope that this is still usable in some areas (timeseries queries?)
3. Pull all of this into an external module, where adherence with normal assumptions and expectations of how things work don't have to be adhered to.

himanshug replied:
ok, let us do (1) optionally , let user specify maxOffheapSize to use.. if sketch fits within that size then it gets stored off-heap (with significant additional serde cost) or else it gets stored on-heap. Now users can decide where they want to get it stored.
we should still keep #2827 so that user can choose to store on-heap but tune topN if/when necessary.

cheddar replied:
That sounds reasonable.  I'm not a fan of #2827 but as long as it defaults to off, I can be ok with it.

himanshug replied:
yes, with #2827 default behavior would still be the same.

fjy commented on docs/content/development/extensions-core/datasketches-aggregators.md:
````diff
@@ -153,3 +162,105 @@ sample query for, How many unique users visited both product A and B?
 }
 ```
+
+## quantiles aggregator ( EXPERIMENTAL )
````
please use the experimental <div> format to highlight this

himanshug replied:
done

fjy commented on docs/content/development/extensions-core/datasketches-aggregators.md:
```diff
@@ -153,3 +162,105 @@ sample query for, How many unique users visited both product A and B?
+  //following field is optional, default = 1024.
+  //it must be a power of 2 and less than 65536
+  //default value here is arbitrary and appropriate value for specific usecase
+  //should be determined. See [quantiles-accuracy](http://datasketches.github.io/docs/QuantilesAccuracy.html) .
```
have you tried rendering these docs with jekyll? I really think the link to the accuracy should be linked outside of a comment in a json blob

himanshug replied:
moved it out in the description above.

fjy commented on docs/content/development/extensions-core/datasketches-aggregators.md:
````diff
@@ -153,3 +162,105 @@ sample query for, How many unique users visited both product A and B?
+  //and you should use same or greater size at query time to get desired accuracy.
+  "size": 1024
+ }
+```
````
can we add some docs about when to use this vs the other histogram extension?

himanshug replied:
placed a note in description

fjy commented:
👍  if comments on docs are addressed. I think high level the impl is fine, however, what is really required is for this extension to be tested and verified in production. We may want to consider moving this to extensions-contrib until it has seen more production use.

fjy commented:
👍

cheddar commented on extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/QuantilePostAggregator.java:
```diff
@@ -0,0 +1,135 @@
+  @JsonCreator
+  public QuantilePostAggregator(
+      @JsonProperty("name") String name,
+      @JsonProperty("fraction") double fraction,
```
Why not call it the "quantile"?

cheddar commented on extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/QuantilesSketchAggregatorFactory.java:
```diff
@@ -0,0 +1,327 @@
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") String fieldName,
+      @JsonProperty("size") Integer size,
+      @JsonProperty("shouldFinalize") Boolean shouldFinalize,
```
What is the purpose of "shouldFinalize"?  If you don't want things finalized you can already just send your query with `"finalize": false` in the context.

fjy commented:
As discussed with @himanshug, we are going to push this back to 0.9.2 to allow more time to implement the off-heap aggregations.

fjy commented:
@himanshug are you still planning on implementing off heap aggregations?

fundead commented:
Quick question about the DataSketch histogram implementation contained in this PR - is this feature still on the roadmap? We've been looking into Druid for analytics and have a req to surface medians and histograms, but are seeing degradation in query times using the current approxHistogram aggregator (20-30s for an uncached query). There's still a bit of investigation to be done on our end around tuning the `resolution` but we're interested in this alternate implementation if it's more accurate and in the single second range for larger datasources.

I see an off-heap implementation is remaining, is this a blocker on using the extension? Has anyone tried running this branch in production?

himanshug commented:
@fundead would you use more of TopN or groupBy? i would expect the results to be more accurate from this implementation.
However, not having off-heap would mean that this will require significant more jvm heap on historicals and some jvm tuning might be necessary to get good performance (not sure how it will compare in terms of runtime performance to existing histogram module).
there are plans to do the off-heap eventually.

fundead commented:
Thanks for the fast reply himanshug - we're currently using TopN for practically all of our queries. From that advice it seems worthwhile for us to look into this implemention, I'll try build from source rebasing this branch over the 0.9.2-rc1 tag and see how it performs.

pdeva commented:
just informing that datasketch library has changed their binary format of quantiles sketch which is incompatible with previous version.
Also version 0.8.3 of datasketch has completely new quantiles api which will cause compile error in current code

himanshug commented:
@pdeva did you use this code and using quantiles aggregators ?

pdeva commented:
@himanshug  not sure what you mean. are you saying if i am using this code in production?

himanshug commented:
@pdeva yes, are you or did you? 
and if you've been using sketches-core quantiles code in production in any incantation ?

pdeva commented:
not yet.
i am writing my own extension to use the latest version of data-sketches lib though. so it should be in production this week.

himanshug commented:
I have updated this PR to use latest version of sketches-core and also moved it completely into a contrib extension as this can only work with java-8. now on, it should have no conflicts and people should be able to try this out.

himanshug commented:
this is being tested internally on some clusters and we will release new code to once some testing is done with off-heap quantiles sketches.

keeping this PR open in the meantime if someone wanted to give it a test ride.

hamlet-lee commented:
I am looking forward to use this extension.
Current "histfold" aggregator in druid seems to have bug and produces wrong result.

gianm commented:
@hamlet-lee which bug are you talking about? The approx. histogram extension is still supported and bug reports are appreciated if you ran into bad behavior!

btw, @himanshug, re:

> I have updated this PR to use latest version of sketches-core and also moved it completely into a contrib extension as this can only work with java-8.

If you haven't already, you could put this into the main datasketches extension now since Druid now requires java 8.

hamlet-lee commented:
@gianm , I doc the bug at #3972

freakyzoidberg commented:
@himanshug is that PR still active? 
Interested in giving it a go, is there any impactful open issue that you are aware of with the current state of the extension?

gianm commented:
@himanshug are you still planning on working on this?

himanshug commented:
@AlexanderSaydakov is working on it now, we're preparing to release a version internally for testing, there will be a separate PR for that.

gianm commented:
ok, thank you.

hamlet-lee commented:
I have rebase this PR on 0.10.0-rc2, and run it for 5 months in our cluster.
It works fine.  No obvious problem found.

my branch: https://github.com/hamlet-lee/druid/tree/mod_0.10.0-rc2
